### ISSUE TITLE
Lr_scheduler step where to use , warning

### ISSUE THREAD

**Issue by msverma101** [Lightning-AI/lightning#14903]
i want to use the the lr scheduler for gan and i am trying to figure out where to give the lr_scheduler_dis.step()
if i add it in training_step after loss is updated i am getting an warning which i want to fix
Warning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate

    def _gen_step(self, real_images, conditioned_images, scalar_gen, scalar_dis, aux_labels):
        # Pix2Pix has adversarial and a reconstruction loss
        # First calculate the adversarial loss
        # gen_opt = self.optimizers()

        fake_images = self.gen(conditioned_images, scalar_gen)
        dis_logits, aux_output = self.dis(fake_images, conditioned_images, scalar_dis)

        adversarial_loss = self.adversarial_criterion(dis_logits, torch.ones_like(dis_logits))
        criterion_loss = self.criterion(aux_output, aux_labels)#labels need to add
        error_fake = adversarial_loss + criterion_loss #adding real loss and criterion loss from labels
        error_fake.backward(retain_graph=True)#here not sure where to place this
        
        self.log("6_error_fake_gen", error_fake)
        self.log("2_Generator_adver_Loss", adversarial_loss)

        # calculate reconstruction loss
        recon_loss = self.recon_criterion(fake_images, real_images)
        lambda_recon = self.lambda_recon
        self.log("3_Generator_recon_Loss", recon_loss)

        gen_loss = adversarial_loss + lambda_recon * recon_loss

        # gen_opt.zero_grad()
        # self.manual_backward(gen_loss)
        # gen_opt.step()

        return gen_loss

    

    def _dis_step(self, real_images, conditioned_images, scalar_gen, scalar_dis, aux_labels):
        # dis_opt = self.optimizers()#turinign of auto opt need to specify opt manually
        fake_images = self.gen(conditioned_images, scalar_gen)
        fake_logits, _ = self.dis(fake_images, conditioned_images, scalar_dis)

        real_logits, aux_output = self.dis(real_images, conditioned_images, scalar_dis)

        fake_loss = self.adversarial_criterion(fake_logits, torch.zeros_like(fake_logits))
        real_loss = self.adversarial_criterion(real_logits, torch.ones_like(real_logits))

        criterion_loss = self.criterion(aux_output, aux_labels)#labels need to add
        error_real = real_loss + criterion_loss #adding real loss and criterion loss from labels
        error_real.backward(retain_graph=True)#here not sure where to place this
        self.log("7_error_real_dis", error_real)

        accuracy = self.compute_acc(aux_output, aux_labels)

        self.log("4_Discriminator_fake_Loss", fake_loss)
        self.log("5_Discriminator_real_Loss", real_loss)
        self.log("7_accruracy", accuracy)

        dis_loss = (real_loss + fake_loss) / 2

        # dis_opt.zero_grad()
        # self.manual_backward(dis_loss)
        # dis_opt.step()
        return dis_loss #, error_real

    def configure_optimizers(self):
        dis_opt = torch.optim.Adam(self.dis.parameters(), lr=self.learning_rate) 
        gen_opt = torch.optim.Adam(self.gen.parameters(), lr=self.learning_rate)
        #learning rate scheduler
        #https://medium.com/mlearning-ai/make-powerful-deep-learning-models-quickly-using-pytorch-lightning-29f040158ef3
        self.lr_scheduler_dis = torch.optim.lr_scheduler.StepLR(dis_opt, step_size  = 10 , gamma = 0.5)
        self.lr_scheduler_gan = torch.optim.lr_scheduler.StepLR(gen_opt, step_size  = 10 , gamma = 0.5)
        
        return dis_opt, gen_opt

_Originally posted by @msverma101 in https://github.com/Lightning-AI/lightning/issues/3051#issuecomment-1259411718_
      

cc @borda @rohitgr7

**Comment by awaelchli** [comment]
Hi
I see you are using manual optimization. You need to call the lr scheduler's step manually, after the optimizer step, but I can't see that in your code. 

Make sure this code is not commented!

```py
    gen_opt.zero_grad()
    self.manual_backward(gen_loss)
    gen_opt.step()
    self.lr_scheduler_gen.step()   # <---- add this
```

And also here:

```py
    dis_opt.zero_grad()
    self.manual_backward(dis_loss)
    dis_opt.step()
    self.lr_scheduler_dis.step()   # <---- add this
```


Another note:
The best practice in Lightning is to return the schedulers from your `configure_optimizers()` like so,

```py
def configure_optimizers(self):
    ....
    return [dis_opt, gen_opt], [lr_scheduler_dis, lr_scheduler_gen]
```


and then access them in your training step:

```py
lr_scheduler_dis, lr_scheduler_gen = `self.lr_schedulers()`
```

[Documentation](https://pytorch-lightning.readthedocs.io/en/stable/model/manual_optimization.html?highlight=manual%20optimization#learning-rate-scheduling)
